src/messages/textDocument_semanticToken.cc

```diff
@@ -0,0 +1,257 @@
+// Copyright 2017-2020 ccls Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#include "indexer.hh"
+#include "message_handler.hh"
+#include "pipeline.hh"
+#include "sema_manager.hh"
+
+#include <clang/Sema/Sema.h>
+#include <algorithm>
+#include <stdexcept>
+#include <string>
+
+MAKE_HASHABLE(ccls::SymbolIdx, t.usr, t.kind);
+
+namespace ccls {
+using namespace clang;
+
+namespace {
+struct SemanticTokens {
+  std::vector<int> data;
+};
+REFLECT_STRUCT(SemanticTokens, data);
+
+struct CclsSemanticHighlightSymbol {
+  int id = 0;
+  SymbolKind parentKind;
+  SymbolKind kind;
+  uint8_t storage;
+  std::vector<std::pair<lsRange, Role>> lsRangeAndRoles;
+};
+
+struct ScanLineEvent {
+  Position pos;
+  Position end_pos; // Second key when there is a tie for insertion events.
+  int id;
+  CclsSemanticHighlightSymbol *symbol;
+  Role role;
+  bool operator<(const ScanLineEvent &o) const {
+    // See the comments below when insertion/deletion events are inserted.
+    if (!(pos == o.pos))
+      return pos < o.pos;
+    if (!(o.end_pos == end_pos))
+      return o.end_pos < end_pos;
+    // This comparison essentially order Macro after non-Macro,
+    // So that macros will not be rendered as Var/Type/...
+    if (symbol->kind != o.symbol->kind)
+      return symbol->kind < o.symbol->kind;
+    // If symbol A and B occupy the same place, we want one to be placed
+    // before the other consistantly.
+    return symbol->id < o.symbol->id;
+  }
+};
+
+}
+
+void MessageHandler::textDocument_semanticTokensFull(
```

Comment:
This replicates too many lines in `message_handler.cc`. You can make the `message_handler.cc` implementation public and call that function.

Reply:
I couldn't find a quote, but as far as I understand `emitSemanticHighlight()` is being replaced by Semantic Tokens.

I remember reading about it in clangd issues, but I cannot find a hard quote on this. Best mention is done [here](https://github.com/clangd/clangd/issues/352#issuecomment-619442147)

I'll remove old semantic highlighting code.